INWX: support changing from null MX to non-null MX

NOTE: Have a general question?  You'll get a better response on the [dnscontrol-discuss](https://groups.google.com/g/dnscontrol-discuss) email list!

**Describe the bug**
With INWX, when changing from null MX to non-null MX records, the ordering matters because INWX enforces that you cannot add further MX records while you have a null MX record

dnscontrol does not seem to handle this, leading to try to add non-null MX records before the null MX record is modified

**To Reproduce**
Steps to reproduce the behavior:
1. Have an INWX domain that has null MX
2. Change to having non-null MX records e.g. the standard set for Google
3. The additions are attempted before the modification of the existing null MX record, causing them to fail

#1: + CREATE MX example.org 10 alt3.aspmx.l.google.com. ttl=3600
FAILURE! (2308) Data management policy violation
#2: + CREATE MX example.org 10 alt4.aspmx.l.google.com. ttl=3600
FAILURE! (2308) Data management policy violation
#3: + CREATE MX example.org 5 alt1.aspmx.l.google.com. ttl=3600
FAILURE! (2308) Data management policy violation
#4: + CREATE MX example.org 5 alt2.aspmx.l.google.com. ttl=3600
FAILURE! (2308) Data management policy violation
...
#6: ± MODIFY MX example.org: (0 . ttl=3600) -> (1 aspmx.l.google.com. ttl=3600)
SUCCESS!

**Expected behavior**
The existing null MX record should be modified first so that the subsequent additions do not fail

**DNS Provider**
INWX

> I'd like to see if its possible to create an integration test that detects this situation.

Seems to work (the test itself at least):
```
$ go test -v -verbose -timeout 0 -provider INWX | grep -i fail
--- FAIL: TestDNSProviders (130.29s)
    --- FAIL: TestDNSProviders/dnscontrol1.com (130.09s)
        --- FAIL: TestDNSProviders/dnscontrol1.com/16:NullMX:unnull (0.12s)
        --- FAIL: TestDNSProviders/dnscontrol1.com/17:NullMXApex:unnull (1.27s)
        --- FAIL: TestDNSProviders/dnscontrol1.com/19:complex_TXT:TXT_with_1_dq-1interior (0.27s)
        --- FAIL: TestDNSProviders/dnscontrol1.com/20:TXT_backslashes:TXT_with_backslashs (0.41s)
FAIL
FAIL    github.com/StackExchange/dnscontrol/v4/integrationTest  130.876s
```

TRANSIP seems to fail the test as well: https://github.com/StackExchange/dnscontrol/actions/runs/7559961364/job/20585007084?pr=2801

If we prefer having some sort of workaround in INWX provider itself, I can try to work something out.

@tlimoncelli Is the CI running basically the integrationTest just like when running them locally? I could then create a new account with INWX with one test domain and add it to the CI pipeline. So that we have this covered as well.

--- a/providers/inwx/inwxProvider.go
+++ b/providers/inwx/inwxProvider.go
@@ -4,10 +4,12 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"slices"
 	"sort"
 	"strings"
 	"time"
 
+	"github.com/fatih/color"
 	"github.com/nrdcg/goinwx"
 	"github.com/pquerna/otp/totp"
 
@@ -228,12 +230,119 @@ func (api *inwxAPI) deleteRecord(RecordID int) error {
 	return api.client.Nameservers.DeleteRecord(RecordID)
 }
 
+// appendDeleteCorrection is a helper function to append delete corrections to the list of corrections
+func (api *inwxAPI) appendDeleteCorrection(corrections []*models.Correction, rec *models.RecordConfig, removals map[string]struct{}) ([]*models.Correction, map[string]struct{}) {
+	// prevent duplicate delete instructions
+	if _, found := removals[rec.ToComparableNoTTL()]; found {
+		return corrections, removals
+	}
+	corrections = append(corrections, &models.Correction{
+		Msg: color.RedString("- DELETE %s %s %s ttl=%d", rec.GetLabelFQDN(), rec.Type, rec.ToComparableNoTTL(), rec.TTL),
+		F: func() error {
+			return api.deleteRecord(rec.Original.(goinwx.NameserverRecord).ID)
+		},
+	})
+	removals[rec.ToComparableNoTTL()] = struct{}{}
+	return corrections, removals
+}
+
+// isNullMX checks if a record is a null MX record.
+func isNullMX(rec *models.RecordConfig) bool {
+	return rec.Type == "MX" && rec.MxPreference == 0 && rec.GetTargetField() == "."
+}
+
+// MXCorrections generates required delete corrections when a MX change can not be applied in an updateRecord call.
+func (api *inwxAPI) MXCorrections(dc *models.DomainConfig, foundRecords models.Records) ([]*models.Correction, models.Records, error) {
+
+	// If a null MX is present in the zone, we have to take special care of any
+	// planned MX changes: No non-null MX records can be added until the null
+	// MX is deleted. If a null MX is planned to be added and the diff is
+	// trying to replace an existing regular MX, we need to delete the existing
+	// MX record because an update would be rejected with "2308 Data management policy violation"
+
+	removals := make(map[string]struct{})
+	corrections := []*models.Correction{}
+	tempRecords := []*models.RecordConfig{}
+
+	// Detect Null MX in foundRecords
+	nullMXInFound := slices.ContainsFunc(foundRecords.GetByType("MX"), isNullMX)
+
+	// Detect Null MX and regular MX in desired records
+	nullMXInDesired := false
+	regularMXInDesired := false
+	for _, rec := range dc.Records.GetByType("MX") {
+		if isNullMX(rec) {
+			nullMXInDesired = true
+		} else {
+			regularMXInDesired = true
+		}
+	}
+
+	// invalid state. Null MX and regular MX are both present in the configuration
+	if nullMXInDesired && regularMXInDesired {
+		return nil, nil, fmt.Errorf("desired configuration contains both Null MX and regular MX records")
+	}
+
+	if nullMXInFound && !nullMXInDesired {
+		// Null MX exists in foundRecords, but desired configuration contains only regular MX records
+		// Safe to delete the Null MX record
+		for _, rec := range foundRecords {
+			if isNullMX(rec) {
+				corrections, removals = api.appendDeleteCorrection(corrections, rec, removals)
+			}
+		}
+	} else if !nullMXInFound && nullMXInDesired {
+		// Null MX is being added, ensure all existing MX records are deleted
+		for _, rec := range foundRecords {
+			if rec.Type == "MX" {
+				corrections, removals = api.appendDeleteCorrection(corrections, rec, removals)
+			}
+		}
+	}
+
+	mxRecords := foundRecords.GetByType("MX")
+	mxonlyDc, err := dc.Copy()
+	if err != nil {
+		return nil, nil, err
+	}
+	mxonlyDc.Records = mxonlyDc.Records.GetByType("MX")
+
+	mxchanges, _, err := diff2.ByRecord(mxRecords, mxonlyDc, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, change := range mxchanges {
+		if change.Type == diff2.CHANGE {
+			// INWX will not apply a MX preference update of >=1 to 0. The updateRecord
+			// endpoint will not report an error, so the zone and config will be out of
+			// sync unless we handle this as a delete then create
+			if change.New[0].MxPreference == 0 && change.Old[0].MxPreference != 0 {
+				corrections, removals = api.appendDeleteCorrection(corrections, change.Old[0], removals)
+			}
+		}
+	}
+
+	// We need to remove the RRs already in corrections
+	for _, rec := range foundRecords {
+		if _, found := removals[rec.ToComparableNoTTL()]; !found {
+			tempRecords = append(tempRecords, rec)
+		}
+	}
+
+	cleanedRecords := models.Records(tempRecords)
+	return corrections, cleanedRecords, nil
+}
+
 // GetZoneRecordsCorrections returns a list of corrections that will turn existing records into dc.Records.
 func (api *inwxAPI) GetZoneRecordsCorrections(dc *models.DomainConfig, foundRecords models.Records) ([]*models.Correction, int, error) {
 
-	var corrections []*models.Correction
+	corrections, records, err := api.MXCorrections(dc, foundRecords)
+	if err != nil {
+		return nil, 0, err
+	}
 
-	changes, actualChangeCount, err := diff2.ByRecord(foundRecords, dc, nil)
+	changes, actualChangeCount, err := diff2.ByRecord(records, dc, nil)
 	if err != nil {
 		return nil, 0, err
 	}

